aspnetcore/fundamentals/configuration.md

```diff
@@ -5,162 +5,981 @@ ms.author: riande
+-->
+<!-- literal_block {"xml:space": "preserve", "language": "c#", "dupnames": [], "linenos": false, "classes": [], "ids": [], "backrefs": [], "source": "/Users/shirhatti/src/Docs/aspnet/fundamentals/configuration/sample/src/UsingOptions/Startup2.cs", "highlight_args": {"linenostart": 1, "hl_lines": [9, 10, 11, 12, 13]}, "names": []} -->
+
+````c#
+
+   public void ConfigureServices(IServiceCollection services)
+   {
+// rm
+       // Adds services required for using options.
+       services.AddOptions();
+
+       // Register the ConfigurationBuilder instance which MyOptions binds against.
+       services.Configure<MyOptions>(Configuration);
+
+       // Registers the following lambda used to configure options.
+       services.Configure<MyOptions>( myOptions =>
+       {
+           myOptions.Option1 = "value1_from_action";
+       });
+
+       // Add framework services.
+       services.AddMvc();
+   }
+
+   ````
+
+You can add multiple configuration providers. Configuration providers are available in NuGet packages. They are applied in order they are registered. Each call to `Configure<TOptions>]` adds an `IConfigureOptions<TOptions>` service to the service container. In the example above, the values of `Option1` and `Option2` are both specified in *appsettings.json*, but the value of `Option1` is overridden by the configured delegate in the highlighted code above. When more than one configuration service is enabled, the last configuration source specified “wins”. With the code above, the `HomeController.Index` method returns `option1 = value1_from_action, option2 = 2`.
+
+When you bind options to configuration, each property in your options type is bound to a configuration key of the form `property[:sub-property:]`. For example, the `MyOptions.Option1` property is bound to the key `Option1`, which is read from the `option1` property in *appsettings.json*. A sub-property sample is shown later in this article.
+
+In the following code, a third `IConfigureOptions<TOptions>` service is added to the service container. It binds `MySubOptions` to the section `subsection` of the *appsettings.json* file:
+
+[!code-csharp[Main](configuration/sample/src/UsingOptions/Startup3.cs?name=snippet1)]
+
+<!-- literal_block {"xml:space": "preserve", "language": "c#", "dupnames": [], "linenos": false, "classes": [], "ids": [], "backrefs": [], "source": "/Users/shirhatti/src/Docs/aspnet/fundamentals/configuration/sample/src/UsingOptions/Startup3.cs", "highlight_args": {"linenostart": 1, "hl_lines": [15, 16]}, "names": []} -->
+
+````c#
+
+   public void ConfigureServices(IServiceCollection services)
+   { // rm
+       // Adds services required for using options.
+       services.AddOptions();
+
+       // Configure with Microsoft.Extensions.Options.ConfigurationExtensions
+       services.Configure<MyOptions>(Configuration);
```

Comment:
As mentioned before, binding the whole configuration should be a very rare occurrence.